### PR TITLE
#506: Student can be graduated

### LINKED ISSUE #506: Student can be graduated

### PR DESCRIPTION
**Pull request by n-gb**
Resolves #506 
screenshot from 2017-07-28 16-22-28

### CODE REVIEW

File: components/courses/app/helpers/admin/courses/season_helper.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    module SeasonHelper
+      def graduate_students_link(season)
+        link_to t('courses.students.graduate'),
+          admin_courses_season_graduate_path(season),
+          method: :post, class: 'ui button green',
+          data: { confirm: t('phrases.confirm') }
+      end
+
+      def can_graduate_students?(season)
+        season.passed? && season.students.where(status: :attending).any?
```
Comment (sveredyuk) [inline components/courses/app/helpers/admin/courses/season_helper.rb]: to policy

File: components/courses/app/controllers/admin/courses/graduates_controller.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+module Admin
+  module Courses
+    class GraduatesController < ::Admin::Courses::BaseController
+      def create
+        current_season.students.where(status: :attending).update_all(status: :graduated)
```
Comment (sveredyuk) [inline components/courses/app/controllers/admin/courses/graduates_controller.rb]: to service

File: components/courses/app/policies/courses/season/graduate_students_policy.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+      def allowed?
+        season.passed? && has_attending_students?
+      end
+
+      private
+
+      attr_reader :season
+
+      def has_attending_students?
+        season.students.where(status: :attending).any?
```
Comment (denys-medynskyi) [inline components/courses/app/policies/courses/season/graduate_students_policy.rb]: `where(status: :attending)` is appearing in 2 places. You can use scope instead.

File: components/courses/spec/policies/courses/graduate/graduate_students_policy_spec.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
+require 'rails_helper'
+
+describe Courses::Season::GraduateStudentsPolicy do
+  let (:season)  { create(:season, title: 'Test Season', status: :passed) }
```
Comment (denys-medynskyi) [inline components/courses/spec/policies/courses/graduate/graduate_students_policy_spec.rb]: 1. You can use `described_class` which is equal to `Courses::Season::GraduateStudentsPolicy`
2. You can use [subject](https://relishapp.com/rspec/rspec-core/v/3-6/docs/subject/explicit-subject) like this:
`subject(:policy) {described_class.new(season)}`

**Review by denys-medynskyi — CHANGES_REQUESTED** [review]
Refactor a few things and everything will be good

File: components/courses/spec/policies/courses/graduate/graduate_students_policy_spec.rb (hunk 4 above)
Comment (n-gb) [reply components/courses/spec/policies/courses/graduate/graduate_students_policy_spec.rb]: Neat!

**Comment by sveredyuk** [comment]
@NadiaGubernator Looks like you rebased this branch with troubles :(

**Comment by n-gb** [comment]
That's not it, pulling the branch from git resulted in conflicts and merge.

**Comment by n-gb** [comment]
I've resolved that trouble, so no more sad smileys, please.

**Comment by sveredyuk** [comment]
@NadiaGubernator conflicts

**Comment by n-gb** [comment]
@sveredyuk done

File: components/courses/app/models/courses/student.rb [hunk 5]
```diff
@@ -25,5 +25,7 @@ class Student < ApplicationRecord
     validates :personal_info,   presence: true
     validates :motivation_info, presence: true
     validates :user_id,         uniqueness: { scope: :season_id }
+
+    scope     :attending, -> { where(status: :attending) }
```
Comment (sveredyuk) [inline components/courses/app/models/courses/student.rb]: remove -> http://api.rubyonrails.org/classes/ActiveRecord/Enum.html

File: components/courses/app/helpers/admin/courses/season_helper.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    module SeasonHelper
+      def graduate_students_link(season)
+        link_to t('courses.students.graduate'),
+          admin_courses_season_graduate_path(season),
+          method: :post, class: 'ui button green',
+          data: { confirm: t('phrases.confirm') }
+      end
+
+      def graduate_policy(season)
+        ::Courses::Season::GraduateStudentsPolicy.new(season)
```
Comment (sveredyuk) [inline components/courses/app/helpers/admin/courses/season_helper.rb]: no policy in view helpers. move at least in controller

File: components/courses/app/views/admin/courses/seasons/show.slim [hunk 7]
```diff
@@ -12,3 +12,4 @@
 
     == render partial: 'chart', locals: {chart: line_chart(stat[:students_by_status]),
             chart_name: t('charts.by_status', resource: t('courses.students.plural'))}
+    = graduate_students_link(season) if graduate_policy(season).allowed?
```
Comment (sveredyuk) [inline components/courses/app/views/admin/courses/seasons/show.slim]: enter before

File: components/courses/spec/features/admin/seasons/show_spec.rb [hunk 8]
```diff
@@ -0,0 +1,33 @@
+require 'rails_helper'
+
+RSpec.describe 'Seasons SHOW' do
+  let!(:season_creator) { ::Courses::Mentor.create(user_id: 1, season_id: 1) }
+
+  context 'season is passed' do
+    let!(:season)    { create(:season, title: 'Test Season', status: :passed) }
+    let!(:student_a) { create(:student, status: :attending, season: season) }
```
Comment (sveredyuk) [inline components/courses/spec/features/admin/seasons/show_spec.rb]: avoid a/b when you could use attending_student or dropped_student

File: components/courses/spec/policies/courses/graduate/graduate_students_policy_spec.rb [hunk 9]
```diff
@@ -0,0 +1,40 @@
+require 'rails_helper'
+
+describe Courses::Season::GraduateStudentsPolicy do
+  let (:season)    { create(:season, title: 'Test Season', status: :passed) }
+  let!(:student)   { create(:student, season: season, status: :attending) }
```
Comment (sveredyuk) [inline components/courses/spec/policies/courses/graduate/graduate_students_policy_spec.rb:5]: let!(:student) too much space { ...

File: components/courses/app/models/courses/student.rb (hunk 5 above)
Comment (n-gb) [reply components/courses/app/models/courses/student.rb]: Are you sure it'll work like this?

Comment (sveredyuk) [reply components/courses/app/models/courses/student.rb]: I am sure that `ActiveRecord::Enum` creates scopes for you for free and you shouldn't duplicate it

Comment (n-gb) [reply components/courses/app/models/courses/student.rb]: I misinterpreted your comment then. Thanks for the explanation

File: components/courses/spec/policies/courses/graduate/graduate_students_policy_spec.rb (hunk 9 above)
Comment (n-gb) [reply components/courses/spec/policies/courses/graduate/graduate_students_policy_spec.rb:5]: It is aligned with subject(:policy) { ...

**Comment by n-gb** [comment]
@sveredyuk review pls :D

File: components/courses/app/controllers/admin/courses/seasons_controller.rb [hunk 10]
```diff
@@ -76,6 +76,10 @@ def stat
         }
       end
 
+      def graduate_policy(season)
+        ::Courses::Season::GraduateStudentsPolicy.new(season)
```
Comment (sveredyuk) [inline components/courses/app/controllers/admin/courses/seasons_controller.rb:80]: ||=

**Review by sveredyuk — APPROVED** [review]

**Pull request merged**